File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def move_dir(y, x, directs):
    diry, dirx = directs
    dy = y + diry
    dx = x + dirx
    if dy < -5 or dy > 5 or dx < -5 or dx > 5:
        return y, x
    return dy, dx

def solution(dirs):
    direction_map = {
        "U": (1, 0),
        "D": (-1, 0),
        "R": (0, 1),
        "L": (0, -1)
    }

    y, x = 0, 0
    checked = set()

    for move in dirs:
        dy, dx = move_dir(y, x, direction_map[move])
        if (dy, dx) == (y, x):
            continue
        edge = tuple(sorted([(y, x), (dy, dx)]))
        checked.add(edge)
        y, x = dy, dx
    return len(checked)

File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function moveDir(y, x, [dy, dx]) {
    const ny = y + dy;
    const nx = x + dx;
    if (ny < -5 || ny > 5 || nx < -5 || nx > 5) {
        return [y, x];
    }
    return [ny, nx];
}

function solution(dirs) {
    const directDict = {"U":[1, 0], "D":[-1,0], "R":[0,1], "L":[0,-1]}
    let y = 0;
    let x = 0;
    let checked = new Set()
    for (let dir of dirs) {
        const [ny,nx] = moveDir(y, x, directDict[dir])
        if (ny === y && nx ===x) continue;
        const edge = [[y, x], [ny, nx]].sort().map(p => p.join(',')).join("|")
        checked.add(edge)
        y=ny
        x=nx
    }
    return checked.size;
}